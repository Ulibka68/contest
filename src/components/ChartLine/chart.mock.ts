export const data = [
  6706,
  7579,
  7798,
  8307,
  7866,
  7736,
  7816,
  7630,
  7536,
  7105,
  7178,
  7619,
  7917,
  7483,
  5772,
  5700,
  5435,
  4837,
  4716,
  4890,
  4753,
  4820,
  4538,
  12162,
  39444,
  25765,
  18012,
  14421,
  13249,
  11310,
  10377,
  9399,
  8917,
  8259,
  7902,
  9442,
  47596,
  36160,
  23866,
  18500,
  15488,
  13722,
  12270,
  13413,
  10574,
  7092,
  7159,
  7880,
  8821,
  8306,
  7780,
  7963,
  7837,
  7611,
  7334,
  7413,
  7015,
  6742,
  6557,
  6593,
  6680,
  6725,
  6345,
  5988,
  6365,
  9911,
  28833,
  19694,
  14873,
  11911,
  10498,
  9708,
  8893,
  8365,
  7960,
  7694,
  45529,
  42858,
  31508,
  23289,
  19147,
  15874,
  14551,
  13124,
  11778,
  10809,
  10522,
  9918,
  9436,
  8617,
  8765,
  8194,
  8035,
  7865,
  7573,
  7422,
  7047,
  7147,
  6861,
  6669,
  6363,
  12073,
  32381,
  21390,
  15311,
  12819,
  11655,
  10696,
  9678,
  9143,
  8296,
  7852,
];
