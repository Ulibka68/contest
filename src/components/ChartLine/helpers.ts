import { IChartView } from "@/components/ChartView";

type LineType = [number, number];
type GraphDataType = LineType[];

export const normalize = (value: number, min: number, max: number): number =>
  Math.abs((value - min) / (max - min));

export const dataToArea = (
  data: number[],
  area: IChartView,
  maxY: number = area.height
): GraphDataType => {
  if (data.length) {
    const xInterval = area.width / data.length;
    return data.map((val, i) => [
      xInterval * i++,
      normalize(val, 0, maxY) * area.height,
    ]);
  }
  return [];
};
