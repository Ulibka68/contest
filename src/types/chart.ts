export type Collection = {
  points: number[];
  name: string;
  color: string;
};

export interface LineChart {
  x: number[];
  maxY: number;
  collection: Collection[];
  togglers: { [key: string]: boolean };
}
