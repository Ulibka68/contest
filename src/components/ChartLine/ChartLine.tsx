import React, { useMemo } from "react";
import { dataToArea } from "./helpers";
import { IChartView } from "@/components/ChartView";

export interface ISVGPath extends React.SVGProps<HTMLOrSVGElement> {
  area: IChartView;
  stroke: string;
  style?: React.CSSProperties;
  strokeWidth: string;
  isActive?: boolean;
  maxY: number;
  data: number[];
}

export const ChartLine: React.FC<ISVGPath> = ({
  area,
  style,
  stroke,
  strokeWidth,
  isActive = false,
  data,
  maxY,
}) => {
  const stylePath = { ...style, opacity: isActive ? "1" : "0" };
  const toSVGCoordinates = useMemo(() => {
    const chartData = dataToArea(data, area, maxY);
    const d = [`M ${chartData[0][0]} ${chartData[0][1]}`];
    const collection = chartData.map((section: number[]): string => {
      const xSection = section[0];
      const ySection = section[1];
      return `L ${xSection} ${ySection}`;
    });
    return d.concat(collection).join(" ");
  }, [data, area, maxY]);
  return (
    <path
      strokeWidth={strokeWidth}
      style={stylePath}
      fill="none"
      stroke={stroke}
      vectorEffect="non-scaling-stroke"
      strokeLinejoin="round"
      d={toSVGCoordinates}
    ></path>
  );
};
