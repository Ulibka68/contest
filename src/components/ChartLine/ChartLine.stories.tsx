import React from "react";
import { ChartLine } from "./ChartLine";
import { data } from "./chart.mock";
import { SVGContainer } from "@/components/SVGContainer";
import { ChartView } from "@/components/ChartView";

export default {
  title: "Example/ChartLine",
  component: ChartLine,
};
const area = { width: 800, height: 400 };
const Template = () => (
  <ChartView area={area}>
    <SVGContainer>
      <ChartLine
        area={area}
        maxY={320000}
        isActive={true}
        stroke={"rgb(75, 217, 100)"}
        strokeWidth="2.2px"
        data={data}
      />
    </SVGContainer>
  </ChartView>
);

export const Primary = Template.bind({});
