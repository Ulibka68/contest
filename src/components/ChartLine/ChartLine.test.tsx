import React from "react";
import { shallow } from "enzyme";
import { ChartLine } from "./ChartLine";
import { data } from "./chart.mock";

describe("ChartLine render", () => {
  const area = { width: 800, height: 400 };
  it("is ChartLine with stable date to match snapshot", () => {
    const wrapper = shallow(
      <ChartLine
        area={area}
        maxY={320000}
        isActive={true}
        stroke={"rgb(75, 217, 100)"}
        strokeWidth="2.2px"
        data={data}
      />
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  it("ChartLine with isActive = true has opacity = 1", () => {
    const wrapper = shallow(
      <ChartLine
        area={area}
        maxY={320000}
        isActive={true}
        stroke={"rgb(75, 217, 100)"}
        strokeWidth="2.2px"
        data={data}
      />
    );
    const path = wrapper.find("path");
    const expectedStyle = { opacity: "1" };
    expect(path.props()).toHaveProperty("style", expectedStyle);
  });

  it("ChartLine without isActive status has opacity = 0", () => {
    const wrapper = shallow(
      <ChartLine
        area={area}
        maxY={320000}
        stroke={"rgb(75, 217, 100)"}
        strokeWidth="2.2px"
        data={data}
      />
    );
    const path = wrapper.find("path");
    const expectedStyle = { opacity: "0" };
    expect(path.props()).toHaveProperty("style", expectedStyle);
  });
});
