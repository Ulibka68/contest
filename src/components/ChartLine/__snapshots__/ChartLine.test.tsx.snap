// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartLine render is ChartLine with stable date to match snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "d": "M 0 8.3825 L 0 8.3825 L 7.142857142857143 9.47375 L 14.285714285714286 9.7475 L 21.42857142857143 10.38375 L 28.571428571428573 9.8325 L 35.714285714285715 9.67 L 42.85714285714286 9.77 L 50 9.5375 L 57.142857142857146 9.42 L 64.28571428571429 8.88125 L 71.42857142857143 8.9725 L 78.57142857142857 9.52375 L 85.71428571428572 9.89625 L 92.85714285714286 9.35375 L 100 7.215000000000001 L 107.14285714285715 7.124999999999999 L 114.28571428571429 6.79375 L 121.42857142857143 6.046250000000001 L 128.57142857142858 5.8950000000000005 L 135.71428571428572 6.1125 L 142.85714285714286 5.94125 L 150 6.0249999999999995 L 157.14285714285714 5.672499999999999 L 164.2857142857143 15.202499999999999 L 171.42857142857144 49.305 L 178.57142857142858 32.20625 L 185.71428571428572 22.515 L 192.85714285714286 18.026249999999997 L 200 16.56125 L 207.14285714285717 14.1375 L 214.2857142857143 12.971250000000001 L 221.42857142857144 11.74875 L 228.57142857142858 11.14625 L 235.71428571428572 10.32375 L 242.85714285714286 9.8775 L 250 11.8025 L 257.14285714285717 59.495 L 264.2857142857143 45.2 L 271.42857142857144 29.8325 L 278.5714285714286 23.125 L 285.7142857142857 19.36 L 292.8571428571429 17.1525 L 300 15.3375 L 307.14285714285717 16.76625 L 314.2857142857143 13.2175 L 321.42857142857144 8.865 L 328.5714285714286 8.94875 L 335.7142857142857 9.85 L 342.8571428571429 11.02625 L 350 10.3825 L 357.14285714285717 9.725 L 364.2857142857143 9.95375 L 371.42857142857144 9.796249999999999 L 378.5714285714286 9.51375 L 385.7142857142857 9.1675 L 392.8571428571429 9.26625 L 400 8.76875 L 407.14285714285717 8.4275 L 414.28571428571433 8.19625 L 421.42857142857144 8.24125 L 428.5714285714286 8.35 L 435.7142857142857 8.40625 L 442.8571428571429 7.9312499999999995 L 450 7.485 L 457.14285714285717 7.95625 L 464.28571428571433 12.38875 L 471.42857142857144 36.041250000000005 L 478.5714285714286 24.6175 L 485.7142857142857 18.591250000000002 L 492.8571428571429 14.88875 L 500 13.1225 L 507.14285714285717 12.135 L 514.2857142857143 11.116249999999999 L 521.4285714285714 10.45625 L 528.5714285714286 9.950000000000001 L 535.7142857142858 9.6175 L 542.8571428571429 56.91125 L 550 53.5725 L 557.1428571428572 39.385 L 564.2857142857143 29.11125 L 571.4285714285714 23.93375 L 578.5714285714286 19.842499999999998 L 585.7142857142858 18.188750000000002 L 592.8571428571429 16.405 L 600 14.7225 L 607.1428571428572 13.51125 L 614.2857142857143 13.1525 L 621.4285714285714 12.3975 L 628.5714285714286 11.795 L 635.7142857142858 10.77125 L 642.8571428571429 10.956249999999999 L 650 10.2425 L 657.1428571428572 10.04375 L 664.2857142857143 9.831249999999999 L 671.4285714285714 9.46625 L 678.5714285714286 9.2775 L 685.7142857142858 8.80875 L 692.8571428571429 8.93375 L 700 8.57625 L 707.1428571428572 8.33625 L 714.2857142857143 7.953749999999999 L 721.4285714285714 15.09125 L 728.5714285714286 40.47625 L 735.7142857142858 26.737499999999997 L 742.8571428571429 19.138749999999998 L 750 16.02375 L 757.1428571428572 14.56875 L 764.2857142857143 13.370000000000001 L 771.4285714285714 12.0975 L 778.5714285714287 11.42875 L 785.7142857142858 10.37 L 792.8571428571429 9.815",
      "fill": "none",
      "stroke": "rgb(75, 217, 100)",
      "stroke-linejoin": "round",
      "stroke-width": "2.2px",
      "style": "opacity:1",
      "vector-effect": "non-scaling-stroke",
    },
    "children": Array [],
    "name": "path",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "d": undefined,
      "fill": undefined,
      "stroke": undefined,
      "stroke-linejoin": undefined,
      "stroke-width": undefined,
      "style": undefined,
      "vector-effect": undefined,
    },
    "x-attribsPrefix": Object {
      "d": undefined,
      "fill": undefined,
      "stroke": undefined,
      "stroke-linejoin": undefined,
      "stroke-width": undefined,
      "style": undefined,
      "vector-effect": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
